salt/transport/tcp.py

```diff
@@ -773,38 +812,200 @@ def handle_future(future):
+
+    def close(self):
+        if self._closing:
+            return
+        self._closing = True
+
+    def __del__(self):
+        self.close()
+
+    def _add_client_present(self, client):
+        id_ = client.id_
+        if id_ in self.present:
+            clients = self.present[id_]
+            clients.add(client)
+        else:
+            self.present[id_] = set([client])
+            if self.presence_events:
+                data = {'new': [id_],
+                        'lost': []}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('change', 'presence')
+                )
+                data = {'present': list(self.present.keys())}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('present', 'presence')
+                )
+
+    def _remove_client_present(self, client):
+        id_ = client.id_
+        if id_ is None or id_ not in self.present:
+            return
+
+        clients = self.present[id_]
+        if client not in clients:
+            return
+
+        clients.remove(client)
+        if len(clients) == 0:
+            del self.present[id_]
+            if self.presence_events:
+                data = {'new': [],
+                        'lost': [id_]}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('change', 'presence')
+                )
+                data = {'present': list(self.present.keys())}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('present', 'presence')
+                )
+
+    @tornado.gen.coroutine
+    def _stream_read(self, client):
+        unpacker = msgpack.Unpacker()
+        while not self._closing:
+            try:
+                client._read_until_future = client.stream.read_bytes(4096, partial=True)
+                wire_bytes = yield client._read_until_future
+                unpacker.feed(wire_bytes)
+                for framed_msg in unpacker:
+                    body = framed_msg['body']
+                    if body['enc'] != 'aes':
+                        # We only accept 'aes' encoded messages for 'id'
+                        continue
+                    crypticle = salt.crypt.Crypticle(self.opts, salt.master.SMaster.secrets['aes']['secret'].value)
+                    load = crypticle.loads(body['load'])
+                    if not self.aes_funcs.verify_minion(load['id'], load['tok']):
+                        continue
+                    client.id_ = load['id']
+                    self._add_client_present(client)
+            except tornado.iostream.StreamClosedError as e:
+                log.debug('tcp stream to {0} closed, unable to recv'.format(client.address))
+                client.close()
+                self._remove_client_present(client)
+                self.clients.discard(client)
+                return
```

Comment:
@jacksontj : I believe that even in Python 2.x, `return` by itself should work and `raise gen.Return()` is not necessary. In Python 2.x, `return` followed by a value will not work, and in this case, `raise gen.Return(some_value)` is required. This is no longer required in Python 3.x.

My quote from my previous post above mentions that, and it doesn't qualify Python 2.x vs Python 3.x, so I believe the quote is referring to both. Also, here is another statement supporting my assertion:

http://stackoverflow.com/questions/27020914/return-from-tornado-gen-engine-function-without-yield


_"return" with no argument is the same as "raise gen.Return()". There may be some confusion about Python 2 and 3: in Python 2 a generator can't return a value, but a bare "return" works the same in Python 2 and 3. Can you say more about what you mean, "return still doesn't work"?_

On the other hand, it doesn't hurt anything to have `raise tornado.gen.Return()`, and it is a valid style choice over `return` if consistency is a goal. I am just saying that it isn't absolutely required.